lib/dal/src/workspace_snapshot/node_weight/component_node_weight.rs

```diff
@@ -120,6 +122,110 @@ impl ComponentNodeWeight {
             EdgeWeightKindDiscriminants::Root,
         ]
     }
+
+    pub(crate) async fn try_upgrade_and_create_external_geometry(
+        ctx: &DalContext,
+        v4_graph: &mut WorkspaceSnapshotGraphV4,
+        default_view_idx: NodeIndex,
+        component_node_weight: &ComponentNodeWeight,
+    ) -> NodeWeightResult<()> {
+        dbg!("upgrade one");
+
+        let component_idx = v4_graph
+            .get_node_index_by_id(component_node_weight.id)
+            .map_err(|e| NodeWeightError::WorkspaceSnapshotGraph(Box::new(e)))?;
+
+        let layer_db = ctx.layer_db();
+        let cas = layer_db.cas();
+
+        let component_content: ComponentContent = cas
+            .try_read_as(&component_node_weight.content_hash())
+            .await?
+            .ok_or_else(|| NodeWeightError::MissingContentFromStore(component_node_weight.id))?;
+
+        // When migrating from graph v3 to v4 all components should be on v1
+        let ComponentContent::V1(content) = component_content.clone() else {
+            return Ok(());
+        };
```

Comment:
Rather than silently pretending that we were able to upgrade this `Component` we should probably return an error. Or at the very least, log an `error!(...)` if we encounter a component content that isn't the correct version.